### ISSUE TITLE
[css-scoping][selectors] Should :where(:host) work?

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#6420]
Currently, the root of a web component can only be selected via `:host` and its varieties (`:host()`, `:host-context()` etc). [[spec](https://drafts.csswg.org/css-scoping/#host-element-in-tree)]

Because `:host` has a specificity of `0,1,0`, it can interfere with UA styles, such as those that set `[hidden] { display: none }`. For that reason, [authors are instructed to explicitly account for that in every web component they write](https://developers.google.com/web/fundamentals/web-components/best-practices#add-a-:host-display-style-that-respects-the-hidden-attribute.) by adding this to the component's stylesheet:

```css
:host([hidden]) { display: none }.
```
Not only is this a hassle for authors, but entirely unrealistic to expect that all authors will fall in line and follow this. In practice, we'll just end up with a lot of components that break `hidden`.
But even beyond the `hidden` attribute, it should be possible to specify host styles that do not override existing UA or author styles.

The whole point of [`:where()`](https://drafts.csswg.org/selectors/#zero-matches) is to rectify exactly that kind of problem, however `:where(:host)` does not currently work. Is there a way to fix that? 

cc @tabatkins

**Comment by emilio** [comment]
I'm confused, how is this related to `:where()` at all? Any author selector will override UA styles, so this already doesn't work:

```html
<!doctype html>
<style>
:where(.foo) {
  display: block;
}
</style>
<div class="foo" hidden>Can you see me?</div>
```

**Comment by tabatkins** [comment]
Yeah, the `hidden` thing is an unfortunate consequence of the fact that the hidden attribute is simply broken, and always has been. There's nothing we can do to fix it, either, unfortunately; in particular, making the change requested here wouldn't do anything for it.

The host element not matching :where() *is* according to spec right now (nothing matches the element unless explicitly specified to do so), but :where() and :is() are definitely special-cases in that they're effectively selector grammar, not selectors themselves. I think it would be reasonable to allow them to match the host element, so long as one of their contained selectors does.

I don't know how UAs currently implement the featureless thing, tho, so I'm not sure if it's a big deal to extend to :where()/:is() or not.

**Comment by LeaVerou** [comment]
Ah, right, brain fart, thanks @emilio.

Still, it would be good if authors were able to style `:host` but have their code be overridable by other author code, so original point still stands.

**Comment by emilio** [comment]
`:where` wouldn't fix that either, since the rules from inside and outside the shadow tree are not sorted by specificity.

What you describe should already be working, as rules from outside the shadow tree go before (non-important) rules from inside the shadow tree:

```html
<!doctype html>
<style>
  [hidden] { display: none }
</style>
<div id="host" hidden>Can you see me?</div>
<script>
  document.getElementById("host").attachShadow({ mode: "open" }).innerHTML = `<style>:host { display: block }</style>`;
</script>
```

**Comment by emilio** [comment]
> I don't know how UAs currently implement the featureless thing, tho, so I'm not sure if it's a big deal to extend to :where()/:is() or not.

It'd be slightly annoying to implement the digging into `:where` / `:is`, specially since it'd be valid to do something like `:where(:host, div)` or something, and that is a bit tricky to reason about...

**Comment by emilio** [comment]
`:host-context` is bad to begin with, see #1914.

**Comment by tabatkins** [comment]
Closing as a dupe of #5093 .

**Comment by tabatkins** [comment]
Never mind, not a dupe, just semantically close to what I was working on. ^_^

**Comment by tabatkins** [comment]
Okay *now* it's fixed, by 066ede420.

(Logical combinations work on featureless elements, but still only match against the explicitly-defined set of selectors.)
